sshtunnel.py

```diff
@@ -512,165 +521,148 @@ def __init__(
-                ssh_private_key or
+            self.ssh_private_key = (
+                self.ssh_private_key or
                 hostname_info.get('identityfile', [None])[0]
             )
-            ssh_port = ssh_port or hostname_info.get('port')
+            self.ssh_port = self.ssh_port or hostname_info.get('port')
             proxycommand = hostname_info.get('proxycommand')
-            ssh_proxy = (
-                ssh_proxy or
-                paramiko.ProxyCommand(proxycommand) if proxycommand else None
+            self.ssh_proxy = (
+                self.ssh_proxy or paramiko.ProxyCommand(proxycommand)
+                if proxycommand else None
             )
         except IOError:
-            self.logger.warning('Could not read SSH configuration file: {0}'
-                                .format(ssh_config_file))
+            self.logger.warning(
+                'Could not read SSH configuration file: {0}'
+                .format(self.ssh_config_file)
+            )
 
-        if not ssh_password:
-            ssh_private_key = paramiko.RSAKey.from_private_key_file(
-                ssh_private_key,
-                password=ssh_private_key_password
-            ) if ssh_private_key else None
+    def _consolidate_settings(self):
+        """Get sure everything is in place"""
+        count = len(self._remote_binds) - len(self._local_binds)
+
+        if count < 0:
+            raise ValueError('Too many local bind addresses '
+                             '(local_bind_addresses > remote_bind_addresses)')
+        self._local_binds.extend([('0.0.0.0', 0) for x in range(count)])
+        if not self.ssh_password:
+            self.ssh_private_key = paramiko.RSAKey.from_private_key_file(
+                self.ssh_private_key,
+                password=self.ssh_private_key_password
+            ) if self.ssh_private_key else None
 
             # Check if a private key was supplied or found in ssh_config
-            if not ssh_private_key:
+            if not self.ssh_private_key:
                 raise ValueError('No password or private key available!')
 
-        if not ssh_port:
-            ssh_port = 22
-
-        check_host(ssh_host)
-        check_port(ssh_port)
-        check_addresses(remote_bind_addresses)
-        check_addresses(local_bind_addresses)
-
-        self._server_list = []
+        if not self.ssh_port:
+            self.ssh_port = 22
 
-        self._ssh_username = ssh_username
-        self._ssh_password = ssh_password
-        self._ssh_host_key = ssh_host_key
-        self._ssh_private_key = ssh_private_key
+    def _raise(self, exception, reason):
+        if self._raise_fwd_exc:
+            raise exception(reason)
 
-        self._threaded = threaded
-
-        self._local_interfaces = self._get_local_interfaces()
-
-        self.logger.info('Connecting to gateway: {0}:{1} as user "{2}".'
-                         .format(ssh_host, ssh_port, ssh_username))
+    def _get_transport(self):
```

Comment:
Better to pass the arguments and receive the result of the helper functions. If a helper function changes the state (working with self), the entire logic becomes much more complicated.

For example:
```
@staticmethod
def _get_transport(ssh_address, ssh_proxy=None, keepalive=0, daemon=DAEMON):
    ...
    return transport
```
easier to understand the logic of the project.